OpenKeychain/src/main/java/org/sufficientlysecure/keychain/keyimport/WebKeyDirectoryClient.java

```diff
@@ -0,0 +1,220 @@
+        }
+    }
+
+    @Override
+    public void add(String armoredKey, ParcelableProxy proxy) throws AddKeyException {
+        throw new UnsupportedOperationException("Uploading keys to Web Key Directory not supported");
+    }
+
+    /**
+     * Web Key Directory returns the entire key even during our searching phase.
+     *
+     * @throws UnsupportedOperationException if the key could not be parsed
+     */
+    @NonNull
+    private static ImportKeysListEntry getEntry(byte[] data)
+            throws UnsupportedOperationException, IOException, PgpGeneralException {
+        ImportKeysListEntry entry = new ImportKeysListEntry();
+        entry.setSecretKey(false); // keys imported from Web Key Directory are public
+
+        UncachedKeyRing ring = UncachedKeyRing.decodeFromData(data);
+        UncachedPublicKey key = ring.getPublicKey();
+
+        entry.setUserIds(key.getUnorderedUserIds());
+        entry.setPrimaryUserId(key.getPrimaryUserIdWithFallback());
+
+        entry.setKeyId(key.getKeyId());
+        entry.setParcelableKeyRing(ParcelableKeyRing.createFromEncodedBytes(data));
+        entry.setFingerprint(key.getFingerprint());
+
+        try {
+            entry.setBitStrength(key.getBitStrength());
+            final int algorithm = key.getAlgorithm();
+            entry.setAlgorithm(KeyFormattingUtils.getAlgorithmInfo(algorithm, key.getBitStrength(),
+                    key.getCurveOid()));
+        } catch (NumberFormatException | NullPointerException e) {
+            Timber.e(e, "Conversion for bit size, algorithm, or creation date failed.");
+            // can't use this key
+            throw new UnsupportedOperationException(
+                    "Conversion for bit size, algorithm, or creation date failed.");
+        }
+
+        return entry;
+    }
+
+    @Nullable
+    private static URL toWebKeyDirectoryURL(String name) {
+        Matcher matcher = EMAIL_PATTERN.matcher(name);
+
+        if (!matcher.matches()) {
+            return null;
+        }
+
+        String domain = matcher.group(2);
+        String localPart = matcher.group(1);
+        String encodedPart = encodeZBase32(toSHA1(localPart.getBytes()));
+
+        try {
+            return new URL("https://" + domain + "/.well-known/openpgpkey/hu/" + encodedPart);
+        } catch (MalformedURLException e) {
+            return null;
+        }
+    }
+
+    private static byte[] toSHA1(byte[] input) {
+        try {
+            return MessageDigest.getInstance("SHA-1").digest(input);
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new AssertionError("SHA-1 should always be available");
+        }
+    }
+
+    private static final char[] ALPHABET = "ybndrfg8ejkmcpqxot1uwisza345h769".toCharArray();
+    private static final int SHIFT = Integer.numberOfTrailingZeros(ALPHABET.length);
+    private static final int MASK = ALPHABET.length - 1;
+
+    private static String encodeZBase32(byte[] data) {
```

Comment:
Please move this into a `util.ZBase32` class, it doesn't belong here. A link to the spec would also be nice